verification/certs.go

```diff
@@ -65,8 +167,399 @@ func getTcbInfoForHandle(c DPEClient, handle *ContextHandle) (*ContextHandle, Di
+		keyUsageNames = append(keyUsageNames, "DataEncipherment")
+	}
+
+	if keyUsage&x509.KeyUsageKeyAgreement != 0 {
+		keyUsageNames = append(keyUsageNames, "KeyAgreement")
+	}
+
+	if keyUsage&x509.KeyUsageCertSign != 0 {
+		keyUsageNames = append(keyUsageNames, "CertSign")
+	}
+
+	if keyUsage&x509.KeyUsageCRLSign != 0 {
+		keyUsageNames = append(keyUsageNames, "CRLSign")
+	}
+
+	if keyUsage&x509.KeyUsageEncipherOnly != 0 {
+		keyUsageNames = append(keyUsageNames, "EncipherOnly")
+	}
+
+	if keyUsage&x509.KeyUsageDecipherOnly != 0 {
+		keyUsageNames = append(keyUsageNames, "DecipherOnly")
+	}
+
+	return keyUsageNames
+}
+
+// Verifies the TCI_Current and TCI_Cumulative of dice tcb information blocks
+func verifyDiceTcbDigest(tcbInfo DiceTcbInfo, wantCurrentTCI []byte, lastCumulativeTCI []byte) error {
+	var err error
+
+	// Check TCI_CURRENT
+	currentTCI := tcbInfo.Fwids[0].Digest
+	if !bytes.Equal(currentTCI, wantCurrentTCI) {
+		err = fmt.Errorf("unexpected TCI_CURRENT digest, want %v but got %v", wantCurrentTCI, currentTCI)
+	}
+
+	// Check TCI_CUMULATIVE against expected cumulative TCI
+	wantCumulativeTCI := computeExpectedCumulative(lastCumulativeTCI, currentTCI)
+	cumulativeTCI := tcbInfo.Fwids[1].Digest
+	if !bytes.Equal(cumulativeTCI, wantCumulativeTCI) {
+		err = fmt.Errorf("unexpected TCI_CUMULATIVE value, want %v but got %v", wantCumulativeTCI, cumulativeTCI)
+	}
+	return err
+}
+
+// Checks the FWID block's Digest.
+// FWID at index 0 has the TCI_CURRENT as digest
+// FWID at index 1 has the TCI_CUMULATIVE as digest
+// FWID array always has two digest/hashAlg blocks when "ExtendTci" is supported by DPE profile.
+func validateDiceTcbFwids(leafCertBytes []byte, currentTcis [][]byte, digestLen int) error {
+	var leafCert *x509.Certificate
+	var err error
+
+	// Check whether certificate is DER encoded.
+	if leafCert, err = x509.ParseCertificate(leafCertBytes); err != nil {
+		return err
+	}
+
+	// Get DICE information from MultiTcbInfo Extension
+	var multiTcbInfo []DiceTcbInfo
+	if multiTcbInfo, err = getMultiTcbInfo(leafCert); err != nil {
+		return err
+	}
+
+	if len(multiTcbInfo) == 0 {
+		return fmt.Errorf("certificate MutliTcbInfo is empty")
+	}
+
+	// Calculate expected cumulative value
+	defaultTci := make([]byte, digestLen)
+
+	// Check cumulative, current TCI at the last index of  multitcb info
+	// It must have default TCI value
+	lastIndex := len(multiTcbInfo) - 1
+	if !bytes.Equal(multiTcbInfo[lastIndex].Fwids[0].Digest, defaultTci) {
+		return fmt.Errorf("current TCI value for first TCB block, want %v but got %v", defaultTci, multiTcbInfo[lastIndex].Fwids[0].Digest)
+	}
+
+	if !bytes.Equal(multiTcbInfo[lastIndex].Fwids[1].Digest, defaultTci) {
+		return fmt.Errorf("cumulative TCI value for first TCB block, want %v but got %v", defaultTci, multiTcbInfo[lastIndex].Fwids[1].Digest)
+	}
+
+	// Check cumulative, current TCI of other indices if any
+	lastCumulativeTCI := defaultTci
+	multiTcbInfo = multiTcbInfo[:lastIndex]
+
```

Comment:
Separate validation for the last index as the TCI_CURRENT = TCI_CUMULATIVE = Default TCI.